File Share ACL create and delete services updating file share status are corrected - Fix #930

Shruthi-1MN opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
Problem statement:
==============
Create file share acl code was updating file share status when it fails to create and deletion of acl as well.
There was not status field in Fileshareaclspec to update it's status when it is successful or fails.
So #794 PR is updating fileshare status when acl creation fails, it was causing fatal error so issue has been raised #930 

Solution/Fix:
==========
Added  status field in Fileshareaclspec to tract it's status of creation and failure.
Whenever ACL creation fails to create, it will be update file share acl status.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

PravinRanjan10 commented:
One question, why we need status for fileshare ACL??. How it solve the #930.

Shruthi-1MN commented:
> One question, why we need status for fileshare ACL??. How it solve the #930.
Please check the old code, if acl doesn't create successfully then your were updating fileshare status. Actually you should update file share acl status. 
if err != nil {
		// Change the status of the file share acl to error when the creation faild
		defer db.UpdateFileShareStatus(ctx, db.C, fileshare.Id, model.FileShareError)
		log.Error("when create file share acl:", err.Error())
		return pb.GenericResponseError(err), err
	}

leonwanghui commented:
@Shruthi-1MN One question is if the acl has no status, why do we need to update the status of fileshare if acl creation fails?

Shruthi-1MN commented:
> @Shruthi-1MN One question is if the acl has no status, why do we need to update the status of fileshare if acl creation fails?
Good question, that what old code was doing so in this PR i have corrected and added status field for acl to update it's status not file share status

leonwanghui commented on testutils/collection/data.go:
```diff
@@ -229,7 +229,7 @@ var (
 		},
 	}
-
+	
```
Please remove the blank spaces.

leonwanghui commented on pkg/model/proto/model.proto:
```diff
@@ -758,6 +758,8 @@ message DeleteFileShareAclOpts {
     // The metadata of the file share, optional.
     map<string, string> metadata = 11;
+    // The status
+    string status = 12;
```
Do we need to persist status when sending gRPC request?

leonwanghui commented on pkg/model/fileshare.go:
```diff
@@ -39,6 +39,10 @@ type FileShareAclSpec struct {
 
+	// The status of the fileshareacl.
+	// One of: "available", "error" etc.
+	Status string `json:"status,omitempty"`
```
If you want to add the Status field, please also change swagger.yaml

Shruthi-1MN replied:
ok i will raise another PR to correct swagger.yaml

leonwanghui commented on pkg/model/proto/model.proto:
```diff
@@ -784,6 +784,8 @@ message CreateFileShareAclOpts {
     // The metadata of the file share, optional.
     map<string, string> metadata = 11;
+    // The status
+    string status = 12;
```
Please remove status field, it's not required here.

leonwanghui commented on pkg/controller/controller.go:
```diff
@@ -987,10 +987,13 @@ func (c *Controller) DeleteFileShareAcl(contx context.Context, opt *pb.DeleteFil
 	}
 	dockInfo, err := db.C.GetDockByPoolId(ctx, fileshare.PoolId)
 	if err != nil {
+		defer db.UpdateFileShareAclStatus(ctx, db.C, fileshare.Id, model.FileShareAclErrorDeleting)
```
It should not be fileshare.Id here, please change it to acl id.

leonwanghui commented on pkg/api/util/db.go:
```diff
@@ -108,19 +108,37 @@ func CreateFileShareAclDBEntry(ctx *c.Context, in *model.FileShareAclSpec) (*mod
 	}
 
-	// Store the fileshare meadata into database.
+	in.Status = model.FileShareAclAvailable
```
Why this status is changed into `available`?

leonwanghui commented on pkg/api/util/db.go:
```diff
@@ -108,19 +108,37 @@ func CreateFileShareAclDBEntry(ctx *c.Context, in *model.FileShareAclSpec) (*mod
 	return db.C.CreateFileShareAcl(ctx, in)
 }
 
 func DeleteFileShareAclDBEntry(ctx *c.Context, in *model.FileShareAclSpec) error {
```
Please also add some test cases for failing to delete invalid file share acl.

leonwanghui commented on pkg/controller/controller.go:
```diff
@@ -970,7 +970,7 @@ func (c *Controller) CreateFileShareAcl(contx context.Context, opt *pb.CreateFil
 	if err != nil {
 		// Change the status of the file share acl to error when the creation faild
-		defer db.UpdateFileShareStatus(ctx, db.C, fileshare.Id, model.FileShareError)
+		defer db.UpdateFileShareAclStatus(ctx, db.C, opt.Id, model.FileShareAclError)
```
You need to make the file share acl status to `available` if everything works fine.

Shruthi-1MN replied:
139 has update status in pkg/api/util/db.go file
